c_src/erl_optics.c

```diff
@@ -511,13 +544,18 @@ static ERL_NIF_TERM eo_optics_poll(
     if (!optics) return ERROR("get_optics");
     if (!poller) return ERROR("poller_allocation_error");
 
-    ctx->env = env;
-    ctx->map = enif_make_new_map(env);
-
+    if (erlang_poller){
+        assert(ctx);
+        ctx->env = env;
+        ctx->map = enif_make_new_map(env);
```

Comment:
Modifying the a global variable like this is a bit dangerous as you don't have any guarantees that this function isn't called concurrently which would lead to some nasty problems. So since we can't do away with the globals right now, you may want to consider protecting the poller code from concurrent calls. Note that optics does this internally but that won't protect you here since you're modifying the globals before calling optics.

What I would recommend is adding a variable to indicate that the poller is running and return with an error if a concurrent call is made. Unfortunately, since we're dealing with concurrency, it's not as simple as just adding a boolean (I can explain the details in person) which means you have to resort to a lock that support the `try_lock` operations. The erlang nif interface might have that or you can use the one in optics: https://github.com/RAttab/optics/blob/master/src/utils/lock.h#L11

Reply:
Note that this is technically a problem in most of your functions here but I think we can reasonably make the assumption that poll is the function most likely to lead to concurrency issues. Also I think optics will protect you from most other potential screw ups :)